### ISSUE TITLE
Is a deleted variable kept while a function is using it?

### ISSUE THREAD

**Issue by hanche** [elves/elvish#1031]
This is in part a question, in part a documentation request. And just possibly, a bug report. An example, to make it clear:
```
⬥ x = $nil
⬥ fn x [a]{ put $x; x = $a }
⬥ del x
⬥ x one
⮕ $nil
⬥ x two
⮕ one
⬥ x three
⮕ two
```
Clearly, the function has access to `$x` even after it has been deleted from the enclosing namespace. (The same trick works in a module.) However, the documentation does not mention this, so I worry that garbage collection may come around and try to reclaim the `x` variable, leading to all sorts of “interesting” behaviour. I assume, however, that the garbage collector will leave it alone so long as it is referenced from the function. But knowing nothing about elvish internals, I simply don't know for sure.

If the above technique is safe, I think it should be documented. And if it is not, I consider that a bug, as it is a potential foot gun.

(I asked this question on Telegram, but received no answer. So I put it here, where it won't disappear from view so fast.)

**Comment by krader1961** [comment]
That is the nature of closures. You don't need to worry about the garbage collector, after the `del x`, unless there is a bug in the Elvish implementation. The var that was captured by the closures has to remain accessible. Even if it is no longer accessible in its original scope.

**Comment by hanche** [comment]
Indeed, it _should_ behave like closures do. One reason for my doubts being that I am more used to closures being functions declared inside the body of another function (and then exported). And the way I used to think of it as working, is that the closure keeps its access to the enclosing namespace(s). What is new here, is that we are modifying the enclosing namespace by deleting the variable the function was accessing.

But now I went back and had another look at the section on [closure semantics](https://elv.sh/ref/language.html#closure-semantics) in the language manual, and the discussion of **upvalues** puts my mind somewhat at rest. I think I'll just assume that it works right, and then come back and file a bug report if I discover otherwise.

But I still think it could be made clearer (more explicit) in the documentation, perhaps using an example similar to mine above.
